google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java

```diff
@@ -1829,9 +1694,85 @@ private SpannerException yieldError(SpannerException e) {
     }
   }
 
-  private static class GrpcStruct extends BaseStruct<List<Object>> {
+  private static class GrpcStruct extends Struct implements Serializable {
+
+    protected final Type type;
+    protected final List<Object> rowData;
+
+    /**
+     * Builds an immutable version of this struct using {@link Struct#newBuilder()} which is used
+     * as a serialization proxy.
+     */
+    private Object writeReplace() {
+      Builder builder = Struct.newBuilder();
+      List<Type.StructField> structFields = getType().getStructFields();
+      for (int i = 0; i < structFields.size(); i++) {
+        Type.StructField field = structFields.get(i);
+        String fieldName = field.getName();
+        Object value = rowData.get(i);
+        Type fieldType = field.getType();
+        switch (fieldType.getCode()) {
+          case BOOL:
+            builder.set(fieldName).to((Boolean) value);
+            break;
+          case INT64:
+            builder.set(fieldName).to((Long) value);
+            break;
+          case FLOAT64:
+            builder.set(fieldName).to((Double) value);
+            break;
+          case STRING:
+            builder.set(fieldName).to((String) value);
+            break;
+          case BYTES:
+            builder.set(fieldName).to((ByteArray) value);
+            break;
+          case TIMESTAMP:
+            builder.set(fieldName).to((Timestamp) value);
+            break;
+          case DATE:
+            builder.set(fieldName).to((Date) value);
+            break;
+          case ARRAY:
+            switch(fieldType.getArrayElementType().getCode()) {
+              case BOOL:
+                builder.set(fieldName).toBoolArray((Iterable<Boolean>) value);
+                break;
+              case INT64:
+                builder.set(fieldName).toInt64Array((Iterable<Long>) value);
+                break;
+              case FLOAT64:
+                builder.set(fieldName).toFloat64Array((Iterable<Double>) value);
+                break;
+              case STRING:
+                builder.set(fieldName).toStringArray((Iterable<String>) value);
+                break;
+              case BYTES:
+                builder.set(fieldName).toBytesArray((Iterable<ByteArray>) value);
+                break;
+              case TIMESTAMP:
+                builder.set(fieldName).toTimestampArray((Iterable<Timestamp>) value);
+                break;
+              case DATE:
+                builder.set(fieldName).toDateArray((Iterable<Date>) value);
+                break;
+              case STRUCT:
+                builder.add(fieldName, fieldType.getStructFields(), (Iterable<Struct>) value);
```

Comment:
Can you verify that this is recursively calling writeReplace on these inner structs as well? Maybe using a debugger.